src/main/java/org/truffleruby/language/globals/GlobalVariableStorage.java

```diff
@@ -26,18 +28,25 @@
 
     private volatile Object value;
 
+    private final Object defaultValue;
     private final DynamicObject getter;
     private final DynamicObject setter;
 
     GlobalVariableStorage(Object value, DynamicObject getter, DynamicObject setter) {
-        this.value = value;
+        this.defaultValue = value;
+        this.value = UNSET_VALUE;
         this.getter = getter;
         this.setter = setter;
         assert (getter != null) == (setter != null);
     }
 
     public Object getValue() {
-        return value;
+        Object currentValue = value;
+        return currentValue == UNSET_VALUE ? defaultValue : currentValue;
+    }
+
+    public boolean isDefined() {
+        return hasHooks() || value != UNSET_VALUE;
```

Comment:
Won't this consider variables such `$LOAD_PATH` not defined because they are defined with GlobalVariables#put (which should probably be renamed `#register` or so) ?